src/kOS.Safe/Compilation/KS/kRISC.tpg

```diff
@@ -152,7 +153,31 @@ remove_stmt    -> REMOVE expr EOI;
 log_stmt       -> LOG expr TO expr EOI;
 break_stmt     -> BREAK EOI;
 preserve_stmt  -> PRESERVE EOI;
-declare_stmt   -> DECLARE ( (PARAMETER IDENTIFIER (COMMA IDENTIFIER)* EOI) | (FUNCTION IDENTIFIER instruction_block EOI?) | (IDENTIFIER TO expr) EOI);
+declare_identifier_clause -> IDENTIFIER (TO|IS) expr EOI;
+declare_parameter_clause  -> PARAMETER IDENTIFIER (COMMA IDENTIFIER)* EOI;
+declare_function_clause   -> FUNCTION IDENTIFIER instruction_block EOI?;
+declare_lock_clause       -> LOCK IDENTIFIER TO expr EOI;
+declare_stmt   -> // declare|global|local are all optional for function and parameter,
+                  // but mandatory for declaring an identifier.  This was the only
+                  // way to get the LL(1) parser to understand that without barfing on
+                  // ambiguity.
+                  (   
+                      declare_parameter_clause |
+                      declare_function_clause |
+                      declare_lock_clause |
+                      (
+                          (
+                              (DECLARE (LOCAL|GLOBAL)?) |
+                              (LOCAL|GLOBAL)
+                          )
+                          (
+                              declare_parameter_clause |
+                              declare_function_clause |
+                              declare_identifier_clause |
+                              declare_lock_clause
+                          )
+                      )
+                  );
```

Comment:
Many of the changes to the Compiler.cs file are a consequence of this.  Functions and locks and parameters were moved down a step in the parse tree to put them under a parent generic "declare" rule.  Thus the parent rule deals with the 'global', 'local' and 'declare' keywords (if you look at the above rules, you can see it's still called a declare_stmt even when the 'declare' keyword is omitted and it just says "local" or "global" or none of the above.).